extensions/data-transfer/portability-data-transfer-generic/src/main/java/org/datatransferproject/datatransfer/generic/GenericImporter.java

```diff
@@ -0,0 +1,172 @@
+
+  static final MediaType JSON = MediaType.parse("application/json");
+
+  public GenericImporter(
+      ContainerSerializer<C, R> containerSerializer,
+      AppCredentials appCredentials,
+      URL endpoint,
+      Monitor monitor) {
+    this.monitor = monitor;
+    this.appCredentials = appCredentials;
+    this.endpoint = endpoint;
+    this.containerSerializer = containerSerializer;
+    configureObjectMapper(om);
+  }
+
+  @VisibleForTesting
+  static void configureObjectMapper(ObjectMapper objectMapper) {
+    // ZonedDateTime and friends
+    objectMapper.registerModule(new JavaTimeModule());
+    // Optional fields
+    objectMapper.registerModule(new Jdk8Module());
+    // ISO timestamps rather than unix epoch seconds
+    objectMapper.configure(SerializationFeature.WRITE_DATES_AS_TIMESTAMPS, false);
+  }
+
+  @Override
+  public ImportResult importItem(
+      UUID jobId,
+      IdempotentImportExecutor idempotentExecutor,
+      TokensAndUrlAuthData initialAuthData,
+      C data)
+      throws Exception {
+    OAuthTokenManager tokenManager =
+        jobTokenManagerMap.computeIfAbsent(
+            jobId,
+            ignored -> new OAuthTokenManager(initialAuthData, appCredentials, client, monitor));
+    for (ImportableData<R> importableData : containerSerializer.apply(data)) {
+      idempotentExecutor.executeAndSwallowIOExceptions(
+          importableData.getIdempotentId(),
+          importableData.getName(),
+          () ->
+              tokenManager.withAuthData(
+                  authData -> importSingleItem(jobId, authData, importableData)));
+    }
+    return new ImportResult(ResultType.OK);
+  }
+
+  boolean parseResponse(Response response) throws IOException, InvalidTokenException {
+    if (response.code() >= 400) {
+      byte[] body = response.body().bytes();
+      ErrorResponse error;
+      try {
+        error = om.readValue(body, ErrorResponse.class);
+      } catch (JsonParseException | JsonMappingException e) {
+        throw new IOException(
+            format(
+                "Unexpected response (%d) '%s'",
+                response.code(), new String(body, StandardCharsets.UTF_8)),
+            e);
+      }
+
+      if (response.code() == 401 && error.getError().equals("invalid_token")) {
+        throw new InvalidTokenException(error.toString(), null);
+      } else {
+        throw new IOException(format("Error (%d) %s", response.code(), error.toString()));
+      }
+    }
+    if (response.code() < 200 || response.code() >= 300) {
+      throw new IOException(format("Unexpected response code (%d)", response.code()));
+    }
+    return true;
+  }
+
+  boolean importSingleItem(UUID jobId, TokensAndUrlAuthData authData, ImportableData<R> dataItem)
+      throws IOException, InvalidTokenException {
+    Request request =
+        new Request.Builder()
+            .url(endpoint)
+            .addHeader("Authorization", format("Bearer %s", authData.getToken()))
```

Comment:
you don't have to address this now. But probably a good practice to extract these common strings to a highlevel constants file that is accessible to all the extensions.